refactor: merging PXE interface and its implementation

benesjan opened the pull request:
There will always be only 1 implementation of the PXE interface so it didn't really make sense to have it. In this PR I merged it with PXEService (i.e. copy over the docs) and I rename all the occurrences of PXE Service just as PXE. Now that there is no JSON RPC PXE server that name was misleading.


## Staring into the abyss

Unfortunately, there is a downside to this PR and that is that I needed to drop the proper `PXE` type from `BaseWallet` as Aztec.js cannot import pxe package due to a circular dependency: `pxe --> bbprover --> aztec.js --> pxe`

For this reason now there is the pxe arg defined as `pxe: any`. This will get resolved once `@aztec/wallet-sdk` package is introduced.

benesjan commented:
* **#17380** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17380?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17325** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17325?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17295** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17295?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17303** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17303?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17303?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#17283** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17283?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -59,7 +55,8 @@ export abstract class BaseWallet implements Wallet {
   protected constructor(
-    protected readonly pxe: PXE,
+    // TODO: We cannot type here pxe because we cannot import that package as that would result in a circular dependency.
+    protected readonly pxe: any,
```
Not sure what do with this. Sad

Thunkar replied:
This is very problematic...any chance that `bbprover` can be cleaned up from aztec.js and use stdlib directly? That should break the cycle

benesjan replied:
As agreed in DMs we will just learn to live this for a bit and resolve this once the @aztec/wallet-sdk package is introduced.

benesjan commented on yarn-project/cli/README.md:
```diff
@@ -20,420 +20,4 @@ bash -i <(curl -s https://install.aztec.network)
 
 ## Usage
 
-To use `aztec-cli`, open a terminal or command prompt and run the `aztec-cli` command followed by the desired command and its options.
```
This was super stale so I just instructed reader to print out `aztec-cli --help`

benesjan commented on yarn-project/pxe/package.json:
```diff
@@ -3,6 +3,7 @@
   "version": "0.1.0",
   "type": "module",
   "exports": {
+    ".": "./dest/index.js",
```
TBH not sure if this blows up some bundles

Grego halp

Thunkar replied:
Never do this, plz. Exports should have a scope. What are you trying to achieve?

benesjan commented on yarn-project/stdlib/src/abi/event_metadata_definition.ts:
```diff
@@ -0,0 +1,8 @@
+import type { AbiType } from './abi.js';
+import type { EventSelector } from './event_selector.js';
+
+export type EventMetadataDefinition = {
```
Moved here from PXE interface as it's imported by both wallet and pxe

benesjan commented on yarn-project/stdlib/src/contract/contract_class_metadata.ts:
```diff
@@ -0,0 +1,8 @@
+import type { ContractArtifact } from '../abi/abi.js';
+import type { ContractClassWithId } from './interfaces/contract_class.js';
+
+export interface ContractClassMetadata {
```
Moved here from PXE interface as it's imported by both wallet and pxe

benesjan commented on yarn-project/stdlib/src/contract/contract_metadata.ts:
```diff
@@ -0,0 +1,7 @@
+import type { ContractInstanceWithAddress } from './interfaces/contract_instance.js';
+
+export interface ContractMetadata {
```
Moved here from PXE interface as it's imported by both wallet and pxe

Thunkar commented on yarn-project/test-wallet/src/wallet/test_wallet.ts:
```diff
@@ -16,7 +16,7 @@ import {
 import type { ExecutionPayload } from '@aztec/entrypoints/payload';
 import { Fq, Fr } from '@aztec/foundation/fields';
-import type { PXE } from '@aztec/pxe';
+import type { PXE } from '@aztec/pxe/server';
```
Even if just a type import, prefer /client/lazy ones to be safe when these can be imported from the browser

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/17303/commits/1737485474847ae3b236561881c44748c8497e7b

benesjan committed:
refactor: merging PXE interface and its implementation

There will always be only 1 implementation of the PXE interface so it didn't really make sense to have it. In this PR I merged it with PXEService (i.e. copy over the docs) and I rename all the occurrences of PXE Service just as PXE. Now that there is no JSON RPC PXE server that name was misleading.

## Staring into the abyss

Unfortunately, there is a downside to this PR and that is that I needed to drop the proper `PXE` type from `BaseWallet` as Aztec.js cannot import pxe package due to a circular dependency: `pxe --> bbprover --> aztec.js --> pxe`

For this reason now there is the pxe arg defined as `pxe: any`. This will get resolved once `@aztec/wallet-sdk` package is introduced.
